From fd34dabe4b35ea06d552c4be20bf21792f5b6b88 Mon Sep 17 00:00:00 2001
From: Louis Popi <theh2o64@gmail.com>
Date: Thu, 20 Oct 2016 19:06:41 +0200
Subject: [PATCH 125/178] input: touchscreen: synaptics_dsx_force: Implement
 swap button interface

 * Add a way to swap BACK and MENU buttons
 * Synaptics DSX touchscreen drivers report KEY_MENU as left button
   and KEY_BACK as right button by default

Change-Id: Iccd57a22e489ad4e1e06b2c53b7ca6d9ed5fb2a9
---
 .../synaptics_dsx_force/synaptics_dsx_core.c  | 61 ++++++++++++++++---
 .../synaptics_dsx_force/synaptics_dsx_core.h  |  1 +
 2 files changed, 53 insertions(+), 9 deletions(-)

diff --git a/drivers/input/touchscreen/synaptics_dsx_force/synaptics_dsx_core.c b/drivers/input/touchscreen/synaptics_dsx_force/synaptics_dsx_core.c
index 4f1695e01f73..3f3a74a3fb34 100644
--- a/drivers/input/touchscreen/synaptics_dsx_force/synaptics_dsx_core.c
+++ b/drivers/input/touchscreen/synaptics_dsx_force/synaptics_dsx_core.c
@@ -224,7 +224,11 @@ static ssize_t synaptics_rmi4_0dbutton_show(struct device *dev,
 
 static ssize_t synaptics_rmi4_0dbutton_store(struct device *dev,
 		struct device_attribute *attr, const char *buf, size_t count);
+static ssize_t synaptics_rmi4_reversed_keys_show(struct device *dev,
+		struct device_attribute *attr, char *buf);
 
+static ssize_t synaptics_rmi4_reversed_keys_store(struct device *dev,
+		struct device_attribute *attr, const char *buf, size_t count);
 static ssize_t synaptics_rmi4_suspend_store(struct device *dev,
 		struct device_attribute *attr, const char *buf, size_t count);
 
@@ -766,6 +770,9 @@ static struct device_attribute attrs[] = {
 	__ATTR(0dbutton, (S_IRUGO | S_IWUSR),
 			synaptics_rmi4_0dbutton_show,
 			synaptics_rmi4_0dbutton_store),
+	__ATTR(reversed_keys, (S_IRUGO | S_IWUSR),
+			synaptics_rmi4_reversed_keys_show,
+			synaptics_rmi4_reversed_keys_store),
 	__ATTR(suspend, S_IWUSR,
 			synaptics_rmi4_show_error,
 			synaptics_rmi4_suspend_store),
@@ -1146,6 +1153,29 @@ static ssize_t synaptics_rmi4_0dbutton_store(struct device *dev,
 	return count;
 }
 
+static ssize_t synaptics_rmi4_reversed_keys_show(struct device *dev,
+		struct device_attribute *attr, char *buf)
+{
+	struct synaptics_rmi4_data *rmi4_data = dev_get_drvdata(dev);
+
+	return snprintf(buf, PAGE_SIZE, "%u\n",
+			rmi4_data->enable_reversed_keys);
+}
+
+static ssize_t synaptics_rmi4_reversed_keys_store(struct device *dev,
+		struct device_attribute *attr, const char *buf, size_t count)
+{
+	unsigned int input;
+	struct synaptics_rmi4_data *rmi4_data = dev_get_drvdata(dev);
+
+	if (sscanf(buf, "%u", &input) != 1)
+		return -EINVAL;
+
+	rmi4_data->enable_reversed_keys = !!input;
+
+	return count;
+}
+
 static ssize_t synaptics_rmi4_suspend_store(struct device *dev,
 		struct device_attribute *attr, const char *buf, size_t count)
 {
@@ -1800,6 +1830,25 @@ static int synaptics_rmi4_f12_abs_report(struct synaptics_rmi4_data *rmi4_data,
 	return touch_count;
 }
 
+static void synaptics_rmi4_report_key(struct synaptics_rmi4_data *rmi4_data,
+				      int key, int status)
+{
+	if (rmi4_data->enable_reversed_keys) {
+		switch (key) {
+		case KEY_MENU:
+			key = KEY_BACK;
+			break;
+		case KEY_BACK:
+			key = KEY_MENU;
+			break;
+		default:
+			break;
+		}
+	}
+
+	input_report_key(rmi4_data->input_dev, key, status);
+}
+
 static void synaptics_rmi4_f1a_report(struct synaptics_rmi4_data *rmi4_data,
 		struct synaptics_rmi4_fn *fhandler)
 {
@@ -1868,16 +1917,12 @@ static void synaptics_rmi4_f1a_report(struct synaptics_rmi4_data *rmi4_data,
 			}
 
 			touch_count++;
-			input_report_key(rmi4_data->input_dev,
-					f1a->button_map[button],
-					status);
+			synaptics_rmi4_report_key(rmi4_data, f1a->button_map[button], status);
 		} else {
 			if (before_2d_status[button] == 1) {
 				before_2d_status[button] = 0;
 				touch_count++;
-				input_report_key(rmi4_data->input_dev,
-						f1a->button_map[button],
-						status);
+				synaptics_rmi4_report_key(rmi4_data, f1a->button_map[button], status);
 			} else {
 				if (status == 1)
 					while_2d_status[button] = 1;
@@ -1887,9 +1932,7 @@ static void synaptics_rmi4_f1a_report(struct synaptics_rmi4_data *rmi4_data,
 		}
 #else
 		touch_count++;
-		input_report_key(rmi4_data->input_dev,
-				f1a->button_map[button],
-				status);
+		synaptics_rmi4_report_key(rmi4_data, f1a->button_map[button], status);
 #endif
 	}
 
diff --git a/drivers/input/touchscreen/synaptics_dsx_force/synaptics_dsx_core.h b/drivers/input/touchscreen/synaptics_dsx_force/synaptics_dsx_core.h
index 020e618da2d2..10dffa343f6b 100644
--- a/drivers/input/touchscreen/synaptics_dsx_force/synaptics_dsx_core.h
+++ b/drivers/input/touchscreen/synaptics_dsx_force/synaptics_dsx_core.h
@@ -400,6 +400,7 @@ struct synaptics_rmi4_data {
 	bool f11_wakeup_gesture;
 	bool f12_wakeup_gesture;
 	bool enable_wakeup_gesture;
+	bool enable_reversed_keys;
 	bool wedge_sensor;
 	bool report_pressure;
 	bool stylus_enable;
-- 
2.49.0

